moebot_bot/util/util.go

```diff
@@ -91,6 +93,149 @@ func FindRoleById(roles []*discordgo.Role, toFind string) *discordgo.Role {
 	return nil
 }
 
+func UpdatePollVotes(poll *db.Poll, session *discordgo.Session) error {
+	message, err := session.ChannelMessage(poll.ChannelId, poll.MessageId)
+	if err != nil {
+		return err
+	}
+	for _, o := range poll.Options {
+		r := getReactionById(message, o.ReactionId)
+		if r != nil {
+			o.Votes = r.Count - 1
+		}
+	}
+	return nil
+}
+
+func getReactionById(message *discordgo.Message, reactionId string) *discordgo.MessageReactions {
+	for _, r := range message.Reactions {
+		if reactionId == r.Emoji.Name {
+			return r
+		}
+	}
+	return nil
+}
+
+func OpenPollMessage(poll *db.Poll, user *discordgo.User) string {
+	message := user.Mention() + " created the poll **" + poll.Title + "**!\n"
+	for _, o := range poll.Options {
+		message += ":" + o.ReactionName + ":  " + o.Description + "\n"
+	}
+	message += "Poll ID: " + strconv.Itoa(poll.Id)
+	return message
+}
+
+func ClosePollMessage(poll *db.Poll, user *discordgo.User) string {
+	var message string
+	if poll.Open {
+		message = user.Mention() + " closed " + UserIdToMention(poll.UserId) + "'s poll **" + poll.Title + "**!\n"
+	} else {
+		message = "Poll **" + poll.Title + "** is already closed!\n"
+	}
+	winners := pollWinners(poll)
+	if len(winners) == 0 || winners[0].Votes == 0 {
+		message += "There are no winners!"
+		return message
+	}
+	if len(winners) > 1 {
+		message += "Tied for first place:\n"
+	} else {
+		message += "Poll winner:\n"
+	}
+	for _, o := range winners {
+		message += ":" + o.ReactionName + ":  " + o.Description + "\n"
+	}
+	message += "With " + strconv.Itoa(winners[0].Votes) + " votes!"
+	return message
+}
+
+func pollWinners(poll *db.Poll) []*db.PollOption {
+	var winningOptions []*db.PollOption
+	maxVotes := 0
+	for _, option := range poll.Options {
+		if option.Votes > maxVotes {
+			maxVotes = option.Votes
+		}
+	}
+
+	for _, option := range poll.Options {
+		if option.Votes == maxVotes {
+			winningOptions = append(winningOptions, option)
+		}
+	}
+
+	return winningOptions
+}
+
+func CreatePollOptions(options []string) []*db.PollOption {
+	//TODO: Move to a database table?
```

Comment:
Hmmm... that's a good idea, especially since that would be a perfect setting to configure on a per-server basis! Certainly can remain a `// TODO` for now though 👍